### ISSUE TITLE
Test Failure: com.ibm.ws.microprofile.metrics.tck.launcher.MetricsClassLoaderTest_10.testClassLoaderUnloads

### ISSUE THREAD

**Issue by fmhwong** [OpenLiberty/open-liberty#27487]
RTC 298694

Failed on Java 2 security builds only.

```
testClassLoaderUnloads:java.io.IOException: 2024-01-24-14:15:26:486 Server returned HTTP response code: 500 for URL: http://localhost:8010/metric-servlet/metricServlet
    at java.base/sun.net.[www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1997)](http://www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1997))
    at java.base/sun.net.[www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1589)](http://www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1589))
    at com.ibm.ws.microprofile.metrics.tck.launcher.MetricsClassLoaderTest.getServlet(MetricsClassLoaderTest.java:101)
    at com.ibm.ws.microprofile.metrics.tck.launcher.MetricsClassLoaderTest.testClassLoaderUnloads(MetricsClassLoaderTest.java:44)
    at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
    at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at componenttest.custom.junit.runner.FATRunner$1.evaluate(FATRunner.java:203)
    at componenttest.custom.junit.runner.FATRunner$2.evaluate(FATRunner.java:363)
    at componenttest.custom.junit.runner.FATRunner.run(FATRunner.java:177)
```

**Comment by fmhwong** [comment]
```
[1/27/24, 17:47:28:390 PST] 0000002e com.ibm.ws.webcontainer.util.ApplicationErrorUtils           E SRVE0777E: Exception thrown by application class 'com.ibm.ws.classloading.internal.AppClassLoader.getClassSpecificProtectionDomainPrivileged:485'
java.lang.NullPointerException: Cannot invoke "java.net.URL.openConnection()" because "resourceUrl" is null
	at com.ibm.ws.classloading.internal.AppClassLoader.getClassSpecificProtectionDomainPrivileged(AppClassLoader.java:485)
	at com.ibm.ws.classloading.internal.AppClassLoader$3.run(AppClassLoader.java:470)
	at com.ibm.ws.classloading.internal.AppClassLoader$3.run(AppClassLoader.java:467)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:748)
	at com.ibm.ws.classloading.internal.AppClassLoader.getClassSpecificProtectionDomain(AppClassLoader.java:467)
	at com.ibm.ws.classloading.internal.AppClassLoader.definePackageAndClass(AppClassLoader.java:414)
	at com.ibm.ws.classloading.internal.AppClassLoader.findClass(AppClassLoader.java:333)
	at com.ibm.ws.classloading.internal.AppClassLoader.findOrDelegateLoadClass(AppClassLoader.java:714)
	at com.ibm.ws.classloading.internal.AppClassLoader.loadClass(AppClassLoader.java:586)
	at com.ibm.ws.classloading.internal.AppClassLoader.findClassCommonLibraryClassLoaders(AppClassLoader.java:730)
	at com.ibm.ws.classloading.internal.AppClassLoader.findClass(AppClassLoader.java:327)
	at com.ibm.ws.classloading.internal.AppClassLoader.findOrDelegateLoadClass(AppClassLoader.java:714)
	at com.ibm.ws.classloading.internal.AppClassLoader.loadClass(AppClassLoader.java:586)
	at com.ibm.ws.classloading.internal.AppClassLoader.loadClass(AppClassLoader.java:553)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:1094)
	at com.ibm.ws.microprofile.metrics.fat.metric.servlet.MetricServlet.doGet(MetricServlet.java:42)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:687)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at com.ibm.ws.webcontainer.servlet.ServletWrapper.service(ServletWrapper.java:1260)
	at com.ibm.ws.webcontainer.servlet.ServletWrapper.handleRequest(ServletWrapper.java:748)
	at com.ibm.ws.webcontainer.servlet.ServletWrapper.handleRequest(ServletWrapper.java:445)
	at com.ibm.ws.webcontainer.filter.WebAppFilterManager.invokeFilters(WebAppFilterManager.java:1361)
	at com.ibm.ws.webcontainer.webapp.WebApp.handleRequest(WebApp.java:5076)
	at com.ibm.ws.webcontainer.osgi.DynamicVirtualHost$2.handleRequest(DynamicVirtualHost.java:328)
	at com.ibm.ws.webcontainer.WebContainer.handleRequest(WebContainer.java:1038)
	at com.ibm.ws.webcontainer.osgi.DynamicVirtualHost$2.run(DynamicVirtualHost.java:293)
	at com.ibm.ws.http.dispatcher.internal.channel.HttpDispatcherLink$TaskWrapper.run(HttpDispatcherLink.java:1260)
	at com.ibm.ws.http.dispatcher.internal.channel.HttpDispatcherLink.wrapHandlerAndExecute(HttpDispatcherLink.java:476)
	at com.ibm.ws.http.dispatcher.internal.channel.HttpDispatcherLink.ready(HttpDispatcherLink.java:435)
	at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.handleDiscrimination(HttpInboundLink.java:569)
	at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.handleNewRequest(HttpInboundLink.java:503)
	at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.processRequest(HttpInboundLink.java:363)
	at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.ready(HttpInboundLink.java:330)
	at com.ibm.ws.tcpchannel.internal.NewConnectionInitialReadCallback.sendToDiscriminators(NewConnectionInitialReadCallback.java:169)
	at com.ibm.ws.tcpchannel.internal.NewConnectionInitialReadCallback.complete(NewConnectionInitialReadCallback.java:77)
	at com.ibm.ws.tcpchannel.internal.WorkQueueManager.requestComplete(WorkQueueManager.java:516)
	at com.ibm.ws.tcpchannel.internal.WorkQueueManager.attemptIO(WorkQueueManager.java:586)
	at com.ibm.ws.tcpchannel.internal.WorkQueueManager.workerRun(WorkQueueManager.java:970)
	at com.ibm.ws.tcpchannel.internal.WorkQueueManager$Worker.run(WorkQueueManager.java:1059)
	at com.ibm.ws.threading.internal.ExecutorServiceImpl$RunnableWrapper.run(ExecutorServiceImpl.java:280)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:858)

```

**Comment by tjwatson** [comment]
Ultimately this is being caused by this permission check:

```
java.security.AccessControlException: Access denied ("org.osgi.framework.ServicePermission" "(service.id=149)" "get")
      at java.base/java.security.AccessController.throwACE(AccessController.java:175)
      at java.base/java.security.AccessController.checkPermissionHelper(AccessController.java:237)
      at java.base/java.security.AccessController.checkPermission(AccessController.java:384)
      at java.base/java.lang.SecurityManager.checkPermission(SecurityManager.java:322)
      at org.eclipse.osgi.internal.serviceregistry.ServiceRegistry.checkGetServicePermission(ServiceRegistry.java:1182)
      at org.eclipse.osgi.internal.serviceregistry.ServiceRegistry.getService(ServiceRegistry.java:541)
      at org.eclipse.osgi.internal.framework.BundleContextImpl.getService(BundleContextImpl.java:660)
      at org.eclipse.osgi.internal.url.URLStreamHandlerProxy$LazyURLStreamHandlerService.get(URLStreamHandlerProxy.java:237)
      at org.eclipse.osgi.internal.url.URLStreamHandlerProxy.getRealHandlerService(URLStreamHandlerProxy.java:209)
      at org.eclipse.osgi.internal.url.URLStreamHandlerProxy.parseURL(URLStreamHandlerProxy.java:145)
      at java.base/java.net.URL.<init>(URL.java:674)
```

This then ends up getting thrown out of URL init with a MalformedURLException which is caught and then null returned.  See:

https://github.com/OpenLiberty/open-liberty/blob/14eb597e8423774a685fb1777a5c012a6fa1d63f/dev/com.ibm.ws.artifact.zip/src/com/ibm/ws/artifact/zip/internal/ZipFileEntry.java#L153-L170

**Comment by tjwatson** [comment]
This was fixed by #29776
